_posts/2017-10-20-totw-77.md

```diff
@@ -0,0 +1,180 @@
+approach is to stay away from these objects: you are allowed to re-assign to
+them, or let them go out of scope, but don’t make any other assumptions about
+their state.
+
+Clang-tidy provides some some static-checking to catch use-after move with the
+[misc-use-after-move](http://clang.llvm.org/extra/clang-tidy/checks/misc-use-after-move.html)
+check. However, static-analysis won't ever be able to catch all of these - be on
+the lookout. Call these out in code review, and avoid them in your own code.
+Stay away from the zombies.
+
+## Wait, `std::move` Doesn’t Move?
+
+Yeah, one other thing to watch for is that a call to `std::move()` isn’t 
+actually a move itself, it’s just a cast to an rvalue-reference. It’s only the 
+use of that reference by a move constructor or move assignment that does the 
+work.
+
+```c++
+std::vector<int> foo = GetSomeInts();
+std::move(foo); // Does nothing.
+// Invokes std::vector<int>’s move-constructor.
+std::vector<int> bar = std::move(foo);
+```
+
+This should almost never happen, and you probably shouldn’t waste a lot of
+mental storage on it. I really only mention it if the connection between
+`std::move()` and a move constructor was confusing you.
+
+## Aaaagh! It’s All Complicated! Why!?!
+
+First: it’s really not so bad. Since we have move operations in the majority of
+our value types (including protobufs), we can do away with all of the
+discussions of "Is this a copy? Is this efficient?" and just rely on name
+counting: two names, a copy. Fewer than that: no copy.
+
+Ignoring the issue of copies, value semantics are clearer and simpler to reason
+about. Consider these two operations:
+
+```c++
+void Foo(std::vector<string>* paths) {
+  ExpandGlob(GenerateGlob(), paths);
+}
+
+std::vector<string> Bar() {
+  std::vector<string> paths;
+  ExpandGlob(GenerateGlob(), &paths);
+  return paths;
+}
+```
+
+Are these the same? What about if there is existing data in `*paths`? How can
+you tell? Value semantics are easier for a reader to reason about than
+input/output parameters, where you need to think about (and document) what
+happens to existing data, and potentially whether there is an pointer ownership
+transfer.
+
+Because of the simpler guarantees about lifetime and usage when dealing with
+values (instead of pointers), it is easier for the compiler’s optimizers to
+operate on code in that style. Well-managed value semantics also minimizes hits
+on the allocator (which is cheap but not free). Once we understand how move
+semantics help rid us of copies, the compiler’s optimizers can better reason
+about object types, lifetimes, virtual dispatch, and a host of other issues that
+help generate more efficient machine code.
+
+Since most utility code is now move-aware, we should stop worrying about copies
+and pointer semantics, and focus on writing simple easy-to-follow code. Please
+make sure you understand the new rules: not all google3 interfaces will be
```

Comment:
drop "google3" - maybe "not all legacy interfaces"

Reply:
Done